autoscaler/autoscaler_builder.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util/clock"
+)
+
+// Scaler determines the number of replicas to run
+type AutoScaler struct {
+	Verbose      bool
+	Clock        clock.Clock
+	ConfigMapKey string
+	PollPeriod   time.Duration
+	K8sClient    K8sClientInterface
+	StopCh       chan struct{}
+	ReadyCh      chan<- struct{} // For testing.
+
+	version       string
+	coreLadder    []int
```

Comment:
don't need the extra ladder array if sort.Search is used: see this https://play.golang.org/p/bAemZoClF5

all you need to do is sort the coreScalerMap by the first pair of the tuple